ethereum/spec/src/main/java/tech/pegasys/teku/spec/signatures/LocalSlashingProtectorConcurrentAccess.java

```diff
@@ -0,0 +1,136 @@
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.io.SyncDataAccessor;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class LocalSlashingProtectorConcurrentAccess implements SlashingProtector {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Map<BLSPublicKey, LocalSlashingProtectionRecord> records =
+      new ConcurrentHashMap<>();
+  private final SyncDataAccessor dataAccessor;
+  private final Path slashingProtectionBaseDir;
+
+  public LocalSlashingProtectorConcurrentAccess(
+      final SyncDataAccessor dataAccessor, final Path slashingProtectionBaseDir) {
+    this.dataAccessor = dataAccessor;
+    this.slashingProtectionBaseDir = slashingProtectionBaseDir;
+  }
+
+  @Override
+  public SafeFuture<Boolean> maySignBlock(
+      final BLSPublicKey validator, final Bytes32 genesisValidatorsRoot, final UInt64 slot) {
+    return SafeFuture.of(
+        () -> {
+          final LocalSlashingProtectionRecord entry = acquireLock(validator, genesisValidatorsRoot);
+          try {
+            return entry.writeSigningRecord(
+                dataAccessor, entry.maySignBlock(genesisValidatorsRoot, slot));
+          } finally {
+            entry.unlock();
+          }
+        });
+  }
+
+  @Override
+  public SafeFuture<Boolean> maySignAttestation(
+      final BLSPublicKey validator,
+      final Bytes32 genesisValidatorsRoot,
+      final UInt64 sourceEpoch,
+      final UInt64 targetEpoch) {
+    return SafeFuture.of(
+        () -> {
+          final LocalSlashingProtectionRecord entry = acquireLock(validator, genesisValidatorsRoot);
+          try {
+            return entry.writeSigningRecord(
+                dataAccessor,
+                entry.maySignAttestation(genesisValidatorsRoot, sourceEpoch, targetEpoch));
+          } finally {
+            entry.unlock();
+          }
+        });
+  }
+
+  @Override
+  public Optional<ValidatorSigningRecord> getSigningRecord(final BLSPublicKey validator) {
+    final Optional<LocalSlashingProtectionRecord> record =
+        Optional.ofNullable(records.get(validator));
+    if (record.isEmpty()) {
+      // not loaded yet, just get from file if available, can create structure on use.
+      return getValidatorSigningRecordFromFile(validator);
+    }
+    try {
+      // unlike other occasions, we're not creating if absent, so only lock if present
+      record.ifPresent(LocalSlashingProtectionRecord::lock);
+      return record.map(LocalSlashingProtectionRecord::getSigningRecord);
+    } finally {
+      record.ifPresent(LocalSlashingProtectionRecord::unlock);
+    }
+  }
+
+  LocalSlashingProtectionRecord acquireLock(
+      final BLSPublicKey validator, final Bytes32 genesisValidatorsRoot) {
+    final LocalSlashingProtectionRecord entry =
+        records.computeIfAbsent(validator, __ -> addRecord(validator, genesisValidatorsRoot));
+    // essential that we take a lock here
+    entry.lock();
```

Comment:
I am thinking how to improve this manual locking to be safer in usage, and my suggestion is to move it out of this method, so how about calling it smth like `getRecord` and do record.lock() just after `getRecord`, so it's in the same method where we do `finally {unlock}`. We have just 2 usages and a bit duplication should not be an issue.

Reply:
the whole point of the `acquireLock` method is to get the object and acquire the lock. this is the same as the previous PR.
There's now an exception which wasn't in the last PR, but i think i'd prefer to leave it here because that's what the method is designed to do (everything will do acquireLock, then finally unlock)

Reply:
I renamed the function, it's more consistent so we can run with that - makes the biggest impact in test cases, so thats fine.